### PR TITLE
Made client persistent

### PR DESCRIPTION
**Pull request by Darkros1245**
This pull request introduces significant changes to the database schema, client implementation, and various modules to support persistent storage using a SQLite database. The most important changes include adding new tables for contacts and nicknames, updating the client structure to use the new database, and refactoring the contact manager and key manager for improved functionality.

Database schema updates:
* Added `Contacts` and `Nicknames` tables in `client/client_db/migrations/20241115092655_init.sql` to store contact and nickname information.

Client implementation updates:
* Refactored the `Client` structure to use `ClientDB` for persistent storage instead of `InMemory`. This includes updating the `register` and `login` methods to connect to the SQLite database and manage account information. [[1]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L1-R54) [[2]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526R66-R108) [[3]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L141-R172) [[4]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L161-R196) [[5]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L187-R257) [[6]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L210-R279) [[7]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526R408-R448)

Contact manager updates:
* Modified the `ContactManager` to handle contacts without requiring device IDs initially, and added methods to update contacts with device IDs later. [[1]](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899R6-R16) [[2]](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899L42-R52) [[3]](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899L109-R115) [[4]](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899L128-R126) [[5]](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899L140-L151)

Encryption module updates:
* Updated the `encrypt` and `decrypt` functions to work with the new `ClientDB` storage type. [[1]](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L4-R4) [[2]](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L43-R43) [[3]](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L113-R130) [[4]](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L220-R228)

Key manager updates:
* Refactored the `KeyManager` to include a default implementation and a new constructor for initializing with specific key IDs. [[1]](diffhunk://#diff-6aeb6daa3d7edbec228374dce024d0044504e6ab4d87b545b39425ca3b4270bfL1-R2) [[2]](diffhunk://#diff-6aeb6daa3d7edbec228374dce024d0044504e6ab4d87b545b39425ca3b4270bfL55-R57) [[3]](diffhunk://#diff-6aeb6daa3d7edbec228374dce024d0044504e6ab4d87b545b39425ca3b4270bfR66-R77) [[4]](diffhunk://#diff-6aeb6daa3d7edbec228374dce024d0044504e6ab4d87b545b39425ca3b4270bfL158) [[5]](diffhunk://#diff-6aeb6daa3d7edbec228374dce024d0044504e6ab4d87b545b39425ca3b4270bfL170-R182)

### CODE REVIEW

File: client/src/client.rs [hunk 1]
```diff
@@ -1,54 +1,57 @@
...
 use libsignal_protocol::{
     message_decrypt, process_prekey_bundle, CiphertextMessage, CiphertextMessageType,
     IdentityKeyPair, SignalProtocolError,
 };
 use prost::Message;
-use rand::{rngs::OsRng, Rng};
 use std::time::{SystemTime, UNIX_EPOCH};
 
-pub struct Client<ST: StorageType, SSA: SignalServerAPI> {
+pub struct Client<T: ClientDB, SSA: SignalServerAPI> {
```
Comment (Laursen79) [inline client/src/client.rs:33]:
```suggestion
pub struct Client<T: ClientDB, U: SignalServerAPI> {
```
og husk andre steder

File: client/src/client.rs [hunk 2]
```diff
@@ -60,22 +63,49 @@ impl<ST: StorageType, SSA: SignalServerAPI> Client<ST, SSA> {
         }
     }
 
+    async fn connect(database_url: &str, create_db: bool) -> Result<SqlitePool> {
```
Comment (Laursen79) [inline client/src/client.rs:66]: måske skal den hedde connect_database eller sådan noget?

**Review by Mikkel-Boje — DISMISSED** [review]

File: client/src/client.rs [hunk 3]
```diff
@@ -184,8 +219,42 @@ impl<ST: StorageType, SSA: SignalServerAPI> Client<ST, SSA> {
...
+        let device = Device::new(pool.clone());
+        let contacts = match device.load_contacts().await {
+            Ok(contacts) => {
+                let mut c = HashMap::new();
+                for contact in contacts {
+                    c.insert(contact.service_id, contact);
+                }
+                Ok(c)
+            }
+            Err(err) => Err(SignalClientError::DatabaseError(format!("{err}"))),
```
Comment (Laursen79) [inline client/src/client.rs:237]: lav en struct
```
#[derive(Debug, Display, Error)]
pub struct DatabaseError(SignalProtocolError)
```
or lav SignalClientError om til
```
SignalClientError::DatabaseError(DatabaseError)
```

File: client/src/client.rs [hunk 4]
```diff
@@ -184,8 +219,42 @@ impl<ST: StorageType, SSA: SignalServerAPI> Client<ST, SSA> {
...
+        }?;
+        let (one_time, signed, kyber) = device
+            .get_key_ids()
+            .await
+            .map_err(|err| SignalClientError::DatabaseError(format!("{err}")))?;
```
Comment (Laursen79) [inline client/src/client.rs:242]: lav en struct
```
#[derive(Debug, Display, Error)]
pub struct DatabaseError(SignalProtocolError)
```
or lav SignalClientError om til
```
SignalClientError::DatabaseError(DatabaseError)
```
så kan du måske skrive
```suggestion
            .map_err(DatabaseError)?;
```

File: client/src/contact_manager.rs [hunk 5]
```diff
@@ -39,19 +38,18 @@ impl ContactManager {
...
     }
 
-    pub fn add_contact(
-        &mut self,
-        service_id: &ServiceId,
-        device_id: DeviceId,
-    ) -> Result<(), String> {
+    pub fn new_with_contacts(contacts: HashMap<ServiceId, Contact>) -> Self {
+        Self { contacts }
+    }
```
Comment (Laursen79) [inline client/src/contact_manager.rs:43]: er det her clone?

File: client/src/main.rs [hunk 6]
```diff
@@ -20,26 +18,30 @@ mod test_utils;
 
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn Error>> {
-    dotenv().map_err(|err| SignalClientError::DotenvError(format!("{err}")))?;
+    dotenv()?;
+    let db_url =
```
Comment (Laursen79) [inline client/src/main.rs:22]: ups

File: client/src/client.rs (hunk 1 above)
Comment (Darkros1245) [reply client/src/client.rs:33]: Done

File: client/src/client.rs (hunk 2 above)
Comment (Darkros1245) [reply client/src/client.rs:66]: Done

File: client/src/storage/database.rs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+            .await
+            .map_err(|err| SignalProtocolError::InvalidArgument(format!("{err}")))
+    }
+
+    async fn mark_kyber_pre_key_used(
+        &mut self,
+        _kyber_prekey_id: KyberPreKeyId,
+    ) -> Result<(), SignalProtocolError> {
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (alexsteffensen) [reply client/src/storage/database.rs:276]: Husk at Last Resort Kyber Prekey ikke skal kaldes med den her funktion

File: client/src/main.rs (hunk 6 above)
Comment (Darkros1245) [reply client/src/main.rs:22]: Done?

Comment (Darkros1245) [reply client/src/main.rs:22]: Done

File: client/src/client.rs [hunk 8]
```diff
@@ -336,6 +405,47 @@ impl<ST: StorageType, SSA: SignalServerAPI> Client<ST, SSA> {
...
     }
+
+    pub async fn add_contact(&mut self, alias: &str, service_id: ServiceId) -> Result<()> {
+        self.contact_manager
+            .add_contact(&service_id)
+            .map_err(|err| SignalClientError::ContactError(err))?;
+        let contact = self
+            .contact_manager
+            .get_contact(&service_id)
+            .map_err(|err| SignalClientError::ContactError(err))?;
```
[the remark this answers is not in the document]
Comment (Darkros1245) [reply client/src/client.rs:416]: Done

File: client/src/storage/database.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Darkros1245) [reply client/src/storage/database.rs:276]: Done

**Review by alexsteffensen — DISMISSED** [review]
apetor thumbs up

**Review by AOsnes — APPROVED** [review]
lgtm 👍🏻

**Pull request merged**